doc/neps/nep-0047-array-api-standard.rst

```diff
@@ -0,0 +1,588 @@
+
+The APIs of each of these libraries are largely similar, but with enough
+differences that it’s quite difficult to write code that works with multiple
+(or all) of these libraries. The array API standard aims to address that
+issue, by specifying an API for the most common ways arrays are constructed
+and used. The proposed API is quite similar to NumPy's API, and deviates mainly
+in places where (a) NumPy made design choices that are inherently not portable
+to other implementations, and (b) where other libraries consistently deviated
+from NumPy on purpose because NumPy's design turned out to have issues or
+unnecessary complexity.
+
+For a longer discussion on the purpose of the array API standard we refer to
+the `Purpose and Scope section of the array API standard <https://data-apis.github.io/array-api/latest/purpose_and_scope.html>`__
+and the two blog posts announcing the formation of the Consortium [1]_ and
+the release of the first draft version of the standard for community review [2]_.
+
+The scope of this NEP includes:
+
+- Adopting the 2021 version of the array API standard
+- Adding a separate namespace, tentatively named ``numpy.array_api``
+- Changes needed/desired outside of the new namespace, for example new dunder
+  methods on the ``ndarray`` object
+- Implementation choices, and differences between functions in the new
+  namespace with those in the main ``numpy`` namespace
+- A new array object conforming to the array API standard
+- Maintenance effort and testing strategy
+- Impact on NumPy's total exposed API surface and on other future and
+  under-discussion design choices
+- Relation to existing and proposed NumPy array protocols
+  (``__array_ufunc__``, ``__array_function__``, ``__array_module__``).
+- Required improvements to existing NumPy functionality
+
+Out of scope for this NEP are:
+
+- Changes in the array API standard itself. Those are likely to come up
+  during review of this NEP, but should be upstreamed as needed and this NEP
+  subsequently updated.
+
+
+Usage and Impact
+----------------
+
+*This section will be fleshed out later, for now we refer to the use cases given
+in* `the array API standard Use Cases section <https://data-apis.github.io/array-api/latest/use_cases.html>`__
+
+In addition to those use cases, the new namespace contains functionality that
+is widely used and supported by many array libraries. As such, it is a good
+set of functions to teach to newcomers to NumPy and recommend as "best
+practice". That contrasts with NumPy's main namespace, which contains many
+functions and objects that have been superceded or we consider mistakes - but
+that we can't remove because of backwards compatibility reasons.
+
+The usage of the ``numpy.array_api`` namespace by downstream libraries is
+intended to enable them to consume multiple kinds of arrays, *without having
+to have a hard dependency on all of those array libraries*:
+
+.. image:: _static/nep-0047-library-dependencies.png
```

Comment:
I updated this figure, it's much more realistic now - shows both the current state and (in a different color) what will be made possible by the array API standard.

Reply:
Nice!

Dask looks a little sad in the current picture -- it only has a connection to NumPy. But in fact, Dask also implement a NumPy array like interface, so would also draw a few more connections:

- From Napari and Xarray to Dask. These work today.
- From Astropy/Scikit-Learn/Scikit-Image/SciPy/Einops to Dask. These could work with the new protocol (although they might make the image too cluttered!)

I'm not sure that Holoviews is a great example, since visualization is strange use-case -- if it needs to rendered via the CPU eventually, there isn't much point in using a GPU and/or distributed array backend. Maybe a library like Qutip or YT would be a better example?

Reply:
Great points, thanks. I used QuTiP, made Dask happier, and I added a "dependency on non-array library" connection type - useful for other libraries that depend on SciPy, scikit-image and Xarray. I used libraries.io to check all dependencies, I think it's accurate now.